File: native/app/src/main/java/com/ubb/album_manager/ui/AlbumViewModel.kt
package com.ubb.album_manager.ui

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ubb.album_manager.domain.Album
import com.ubb.album_manager.service.AlbumService
import kotlinx.coroutines.launch

class AlbumViewModel(private val albumService: AlbumService) : ViewModel() {
    private val _albums = MutableLiveData<List<Album>>()
    val albums: LiveData<List<Album>> get() = _albums

    init {
        loadAlbums()
    }

    private fun loadAlbums() {
        viewModelScope.launch {
            albumService.getAll().collect { albumList ->
                _albums.postValue(albumList)
            }
        }
    }

    fun add(album: Album) {
        viewModelScope.launch {
            if (albumService.add(album)) {
                Log.d("AlbumViewModel", "Album added: $album")
            } else {
                Log.e("AlbumViewModel", "Error adding album")
            }
        }
    }

    fun edit(updatedAlbum: Album) {
        viewModelScope.launch {
            if (albumService.edit(updatedAlbum)) {
                Log.d("AlbumViewModel", "Album edited: $updatedAlbum")
            } else {
                Log.e("AlbumViewModel", "Error updating album")
            }
        }
    }

    fun delete(albumId: Int) {
        viewModelScope.launch {
            if (albumService.delete(albumId)) {
                Log.d("AlbumViewModel", "Album deleted was number $albumId")
            } else {
                Log.e("AlbumViewModel", "Error deleting album")
            }
        }
    }
}

File: non-native/lib/service/album_view_model.dart
import 'package:album_manager/service/repository.dart';
import 'package:flutter/material.dart' show ChangeNotifier;
import '../model/album.dart';

class AlbumViewModel with ChangeNotifier {
  final List<Album> _albums = List.from(AlbumRepository.getAll());

  List<Album> get albums => List.unmodifiable(_albums);

  void add(Album album) {
    AlbumRepository.add(album);
    _albums.add(album);
    notifyListeners();
  }

  void edit(Album updatedAlbum) {
    if (AlbumRepository.edit(updatedAlbum.id, updatedAlbum)) {
      final index = _albums.indexWhere((album) => album.id == updatedAlbum.id);
      _albums[index] = updatedAlbum;
      notifyListeners();
    }
  }

  void delete(String albumId) {
    if (AlbumRepository.delete(albumId)) {
      _albums.removeWhere((album) => album.id == albumId);
      notifyListeners();
    }
  }

  Album? findById(String albumId) {
    return _albums.firstWhere((album) => album.id == albumId);
  }
}
